File: backend/app/services/stt/rtzr_client.py
"""
RTZR WebSocket STT 클라이언트 모듈
"""
import asyncio
import websockets
import json
import time
import requests
from queue import Queue
from typing import Optional, Callable
from app.core.logger import setup_logger
from app.config import settings

logger = setup_logger(__name__)


class RTZRWebSocketClient:
    """RTZR WebSocket STT 클라이언트"""

    def __init__(self, room_id: str):
        self.client_id = settings.RTZR_CLIENT_ID
        self.client_secret = settings.RTZR_CLIENT_SECRET
        self.api_base = settings.RTZR_API_URL
        self.room_id = room_id
        self._token = None
        self._sess = requests.Session()
        self.ws = None
        self.audio_queue = Queue()
        self.is_running = False
        self.on_transcript = None
        self.connection_ready = False
        self.pending_audio = []  # WebSocket 연결 전 임시 버퍼

    def get_token(self) -> Optional[str]:
        """토큰 발급"""
        if self._token is None or self._token.get("expire_at", 0) < time.time():
            try:
                resp = self._sess.post(
                    f"{self.api_base}/v1/authenticate",
                    data={
                        "client_id": self.client_id,
                        "client_secret": self.client_secret
                    }
                )

                if resp.status_code != 200:
                    logger.error(f"토큰 발급 실패: {resp.status_code}")
                    return None

                resp.raise_for_status()
                token_data = resp.json()

                self._token = {
                    "access_token": token_data["access_token"],
                    "expire_at": time.time() + 86400
                }
                logger.info(f"[{self.room_id}] ✅ RTZR 토큰 발급 성공")
                return self._token["access_token"]

            except Exception as e:
                logger.error(f"[{self.room_id}] ❌ 토큰 발급 실패: {e}")
                return None

        return self._token["access_token"]

    async def connect_websocket(self, on_transcript: Callable[[str], None]):
        """WebSocket 연결 및 스트리밍"""
        self.on_transcript = on_transcript
        token = self.get_token()

        if not token:
            logger.error(f"[{self.room_id}] 토큰을 가져올 수 없습니다")
            return

        # WebSocket URL 구성
        config = {
            "sample_rate": "16000",
            "encoding": "LINEAR16",
            "use_itn": "true",
            "use_disfluency_filter": "true",
            "use_profanity_filter": "false",
            "use_punctuation": "true",
            "use_word_timestamp": "false"
        }

        params = "&".join([f"{k}={v}" for k, v in config.items()])
        ws_url = f"wss://openapi.vito.ai/v1/transcribe:streaming?{params}"

        headers = {
            "Authorization": f"bearer {token}"
        }

        try:
            async with websockets.connect(ws_url, extra_headers=headers) as websocket:
                self.ws = websocket
                self.is_running = True
                logger.info(f"[{self.room_id}] ✅ STT WebSocket 연결됨")

                # 잠시 대기 후 연결 준비 완료 표시
                await asyncio.sleep(0.5)
                self.connection_ready = True

                # 연결 완료 후 pending_audio 처리
                if self.pending_audio:
                    logger.info(f"[{self.room_id}] 대기 중이던 오디오 {len(self.pending_audio)}개 처리 시작")
                    for audio_data in self.pending_audio:
                        self.audio_queue.put(audio_data)
                    self.pending_audio.clear()

                # 수신 및 송신 태스크 동시 실행
                receive_task = asyncio.create_task(self.receive_messages())
                send_task = asyncio.create_task(self.send_audio())

                await asyncio.gather(receive_task, send_task)

        except Exception as e:
            logger.error(f"[{self.room_id}] ❌ WebSocket 연결 오류: {e}")
        finally:
            self.is_running = False
            self.connection_ready = False

    async def receive_messages(self):
        """WebSocket 메시지 수신"""
        try:
            while self.is_running and self.ws:
                message = await self.ws.recv()
                data = json.loads(message)

                # STT 결과 처리
                if data.get("alternatives"):
                    text = data["alternatives"][0].get("text", "").strip()
                    if text and data.get("final"):
                        # logger.debug(f"[{self.room_id}] 인식: {text}")
                        if self.on_transcript:
                            self.on_transcript(text)

                # 에러 메시지 확인
                if data.get("error"):
                    logger.error(f"[{self.room_id}] ❌ STT 에러: {data.get('error')}")

        except websockets.exceptions.ConnectionClosed:
            pass
        except Exception as e:
            logger.error(f"[{self.room_id}] 메시지 수신 오류: {e}")
        finally:
            self.is_running = False

    async def send_audio(self):
        """오디오 데이터 전송"""
        try:
            sent_count = 0
            empty_count = 0

            while self.is_running and self.ws:
                if not self.audio_queue.empty():
                    audio_data = self.audio_queue.get()
                    await self.ws.send(audio_data)
                    sent_count += 1
                    empty_count = 0

                    # 주기적으로 상태 표시
                    # if sent_count % 100 == 0:
                    #     logger.debug(f"[{self.room_id}] 오디오 처리 중...")
                else:
                    await asyncio.sleep(0.01)
                    empty_count += 1

        except Exception as e:
            logger.error(f"[{self.room_id}] 오디오 전송 오류: {e}")
        finally:
            self.is_running = False

    def add_audio(self, audio_data: bytes):
        """오디오 데이터를 큐에 추가"""
        if self.connection_ready:
            self.audio_queue.put(audio_data)
        else:
            # WebSocket 연결 전이면 pending_audio에 저장
            self.pending_audio.append(audio_data)
            # if len(self.pending_audio) % 10 == 0:  # 10개마다 로그
            #     logger.debug(f"[{self.room_id}] Buffering audio while connecting... ({len(self.pending_audio)} chunks)")

    def disconnect(self):
        """연결 종료"""
        self.is_running = False
        self.connection_ready = False
        if self.ws:
            asyncio.create_task(self.ws.close())

File: backend/src/modules/stt/rtzr-client.ts
import WebSocket from "ws";
import axios from "axios";
import { STTProvider } from "./stt-provider.interface";

interface RTZRConfig {
  clientId: string;
  clientSecret: string;
  apiUrl: string;
}

interface RTZRToken {
  accessToken: string;
  expiresAt: number;
}

export class RTZRClient extends STTProvider {
  private config: RTZRConfig;
  private token: RTZRToken | null = null;
  private ws: WebSocket | null = null;
  private pendingAudioBuffer: Buffer[] = [];
  private reconnectAttempts: number = 0;
  private maxReconnectAttempts: number = 5;
  private isManualDisconnect: boolean = false;
  private lastAudioTime: number = 0;
  private keepAliveInterval: NodeJS.Timeout | null = null;
  private firstAudioSentTime: number = 0;
  private audioChunkCount: number = 0;

  constructor(roomId: string, config: RTZRConfig) {
    super(roomId);
    this.config = config;
  }

  // Get access token from RTZR API
  private async getToken(): Promise<string | null> {
    try {
      // Check if token is still valid
      if (this.token && this.token.expiresAt > Date.now()) {
        return this.token.accessToken;
      }

      // Request new token - RTZR uses form data, not JSON
      const params = new URLSearchParams();
      params.append("client_id", this.config.clientId);
      params.append("client_secret", this.config.clientSecret);

      const response = await axios.post(
        `${this.config.apiUrl}/v1/authenticate`,
        params,
        {
          headers: {
            "Content-Type": "application/x-www-form-urlencoded",
          },
        }
      );

      if (response.status !== 200) {
        console.error(`[STT] Failed to get token: ${response.status}`);
        return null;
      }

      // Store token with expiration
      this.token = {
        accessToken: response.data.access_token,
        expiresAt: Date.now() + 23 * 60 * 60 * 1000, // 23 hours
      };

      return this.token.accessToken;
    } catch (error) {
      console.error(`[STT][${this.roomId}] Failed to get token:`, error);
      return null;
    }
  }

  // Connect to WebSocket
  async connect(): Promise<void> {
    // Reset manual disconnect flag when reconnecting
    this.isManualDisconnect = false;

    const token = await this.getToken();
    if (!token) {
      throw new Error("Failed to obtain access token");
    }

    // WebSocket connection parameters - WHISPER for better accuracy
    const params = new URLSearchParams({
      sample_rate: "24000",
      encoding: "LINEAR16",
      // model_name: "whisper",
      language: "ko",
      use_itn: "true",
      use_disfluency_filter: "true",
      use_profanity_filter: "false",
      use_punctuation: "true",
    });

    const wsUrl = `wss://openapi.vito.ai/v1/transcribe:streaming?${params.toString()}`;

    // Create WebSocket connection
    this.ws = new WebSocket(wsUrl, {
      headers: {
        Authorization: `bearer ${token}`,
      },
    });

    // WebSocket event handlers
    this.ws.on("open", () => {
      this.isConnected = true;
      this.reconnectAttempts = 0;
      this.lastAudioTime = Date.now();
      this.emit("connected");
      this.processPendingAudio();
      this.startKeepAlive();
    });

    this.ws.on("message", (data: WebSocket.Data) => {
      try {
        this.handleMessage(JSON.parse(data.toString()));
      } catch (error) {
        console.error(`[STT][${this.roomId}] Parse error:`, error);
      }
    });

    this.ws.on("error", (error) => {
      console.error(`[STT][${this.roomId}] WebSocket error:`, error);
      this.emit("error", error);
    });

    this.ws.on("close", (code, reason) => {
      this.isConnected = false;
      this.ws = null;
      this.emit("disconnected");
      this.stopKeepAlive();

      // Auto-reconnect
      if (
        !this.isManualDisconnect &&
        this.reconnectAttempts < this.maxReconnectAttempts
      ) {
        this.reconnectAttempts++;
        setTimeout(() => this.connect(), 2000);
      }
    });
  }

  // Handle incoming messages - OPTIMIZED with latency tracking
  private handleMessage(message: any): void {
    if (message.error) {
      console.error(`[STT][${this.roomId}] Error:`, message.error);
      this.emit("error", message.error);
      return;
    }

    if (message.alternatives && message.alternatives.length > 0) {
      const text = message.alternatives[0].text?.trim();
      if (text) {
        // Calculate latency from first audio to first result
        if (this.firstAudioSentTime > 0) {
          const latency = Date.now() - this.firstAudioSentTime;
          const isFinal = message.final || false;

          // Reset after final result
          if (isFinal) {
            this.firstAudioSentTime = 0;
            this.audioChunkCount = 0;
          }
        }

        this.emit("transcript", {
          text,
          confidence: message.alternatives[0].confidence,
          final: message.final || false,
        });
      }
    }
  }

  // Send audio data - FAST PATH with timing
  sendAudio(audioData: Buffer): void {
    if (this.isConnected && this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send(audioData);
      this.lastAudioTime = Date.now();

      // Track first audio sent for latency measurement
      this.audioChunkCount++;
      if (this.audioChunkCount === 1) {
        this.firstAudioSentTime = Date.now();
      } 
    } else {
      if (this.pendingAudioBuffer.length === 0) {
        console.warn(
          `[STT][${this.roomId}] ⚠️  WebSocket not ready, buffering audio. ` +
          `Connected: ${this.isConnected}, WS state: ${this.ws?.readyState}`
        );
      }
      this.pendingAudioBuffer.push(audioData);
      if (this.pendingAudioBuffer.length > 100) {
        this.pendingAudioBuffer.shift();
      }
    }
  }

  // Start keep-alive to prevent stream timeout
  private startKeepAlive(): void {
    this.stopKeepAlive();
    this.keepAliveInterval = setInterval(() => {
      if (Date.now() - this.lastAudioTime > 3000) {
        if (this.ws && this.ws.readyState === WebSocket.OPEN) {
          this.ws.send(Buffer.alloc(4096, 0));
          this.lastAudioTime = Date.now();
        }
      }
    }, 2000);
  }

  // Stop keep-alive
  private stopKeepAlive(): void {
    if (this.keepAliveInterval) {
      clearInterval(this.keepAliveInterval);
      this.keepAliveInterval = null;
    }
  }

  // Process pending audio buffer
  private processPendingAudio(): void {
    if (this.pendingAudioBuffer.length > 0) {
      for (const audio of this.pendingAudioBuffer) {
        if (this.ws && this.ws.readyState === WebSocket.OPEN) {
          this.ws.send(audio);
        }
      }
      this.pendingAudioBuffer = [];
    }
  }

  // End of stream
  endStream(): void {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send("EOS");
    }
  }

  // Disconnect
  disconnect(): void {
    this.isManualDisconnect = true;
    this.stopKeepAlive();

    if (this.ws) {
      this.ws.removeAllListeners();
      try {
        this.endStream();
        this.ws.close();
      } catch (error) {
        console.error(`[STT][${this.roomId}] Close error:`, error);
      }
      this.ws = null;
      this.isConnected = false;
    }

    this.pendingAudioBuffer = [];
    this.reconnectAttempts = 0;
  }

  // Check connection status
  isActive(): boolean {
    return (
      this.isConnected &&
      this.ws !== null &&
      this.ws.readyState === WebSocket.OPEN
    );
  }

  // Get provider name
  getProviderName(): string {
    return "rtzr-vito";
  }
}
